File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-mktenv-5-4/YieldCurve.cpp
// YieldCurve.cpp 
#include "YieldCurve.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

YieldCurve::YieldCurve(TiXmlNode* xmlNode)
: PricingStructure(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //algorithmNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* algorithmNode = xmlNode->FirstChildElement("algorithm");

   if(algorithmNode){algorithmIsNull_ = false;}
   else{algorithmIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- algorithmNode , address : " << algorithmNode << std::endl;
   #endif
   if(algorithmNode)
   {
      if(algorithmNode->Attribute("href") || algorithmNode->Attribute("id"))
      {
          if(algorithmNode->Attribute("id"))
          {
             algorithmIDRef_ = algorithmNode->Attribute("id");
             algorithm_ = boost::shared_ptr<XsdTypeString>(new XsdTypeString(algorithmNode));
             algorithm_->setID(algorithmIDRef_);
             IDManager::instance().setID(algorithmIDRef_,algorithm_);
          }
          else if(algorithmNode->Attribute("href")) { algorithmIDRef_ = algorithmNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { algorithm_ = boost::shared_ptr<XsdTypeString>(new XsdTypeString(algorithmNode));}
   }

   //forecastRateIndexNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* forecastRateIndexNode = xmlNode->FirstChildElement("forecastRateIndex");

   if(forecastRateIndexNode){forecastRateIndexIsNull_ = false;}
   else{forecastRateIndexIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- forecastRateIndexNode , address : " << forecastRateIndexNode << std::endl;
   #endif
   if(forecastRateIndexNode)
   {
      if(forecastRateIndexNode->Attribute("href") || forecastRateIndexNode->Attribute("id"))
      {
          if(forecastRateIndexNode->Attribute("id"))
          {
             forecastRateIndexIDRef_ = forecastRateIndexNode->Attribute("id");
             forecastRateIndex_ = boost::shared_ptr<ForecastRateIndex>(new ForecastRateIndex(forecastRateIndexNode));
             forecastRateIndex_->setID(forecastRateIndexIDRef_);
             IDManager::instance().setID(forecastRateIndexIDRef_,forecastRateIndex_);
          }
          else if(forecastRateIndexNode->Attribute("href")) { forecastRateIndexIDRef_ = forecastRateIndexNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { forecastRateIndex_ = boost::shared_ptr<ForecastRateIndex>(new ForecastRateIndex(forecastRateIndexNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<XsdTypeString> YieldCurve::getAlgorithm()
{
   if(!this->algorithmIsNull_){
        if(algorithmIDRef_ != ""){
             return boost::shared_static_cast<XsdTypeString>(IDManager::instance().getID(algorithmIDRef_));
        }else{
             return this->algorithm_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<XsdTypeString>();
   }
}
boost::shared_ptr<ForecastRateIndex> YieldCurve::getForecastRateIndex()
{
   if(!this->forecastRateIndexIsNull_){
        if(forecastRateIndexIDRef_ != ""){
             return boost::shared_static_cast<ForecastRateIndex>(IDManager::instance().getID(forecastRateIndexIDRef_));
        }else{
             return this->forecastRateIndex_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<ForecastRateIndex>();
   }
}
}

File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-mktenv-5-4/YieldCurve.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class YieldCurve : PricingStructure
    {
        public YieldCurve(XmlNode xmlNode)
        : base(xmlNode)
        {
            XmlNode algorithmNode = xmlNode.SelectSingleNode("algorithm");
            
            if (algorithmNode != null)
            {
                if (algorithmNode.Attributes["href"] != null || algorithmNode.Attributes["id"] != null) 
                {
                    if (algorithmNode.Attributes["id"] != null) 
                    {
                        algorithmIDRef_ = algorithmNode.Attributes["id"].Value;
                        XsdTypeString ob = new XsdTypeString(algorithmNode);
                        IDManager.SetID(algorithmIDRef_, ob);
                    }
                    else if (algorithmNode.Attributes["href"] != null)
                    {
                        algorithmIDRef_ = algorithmNode.Attributes["href"].Value;
                    }
                    else
                    {
                        algorithm_ = new XsdTypeString(algorithmNode);
                    }
                }
                else
                {
                    algorithm_ = new XsdTypeString(algorithmNode);
                }
            }
            
        
            XmlNode forecastRateIndexNode = xmlNode.SelectSingleNode("forecastRateIndex");
            
            if (forecastRateIndexNode != null)
            {
                if (forecastRateIndexNode.Attributes["href"] != null || forecastRateIndexNode.Attributes["id"] != null) 
                {
                    if (forecastRateIndexNode.Attributes["id"] != null) 
                    {
                        forecastRateIndexIDRef_ = forecastRateIndexNode.Attributes["id"].Value;
                        ForecastRateIndex ob = new ForecastRateIndex(forecastRateIndexNode);
                        IDManager.SetID(forecastRateIndexIDRef_, ob);
                    }
                    else if (forecastRateIndexNode.Attributes["href"] != null)
                    {
                        forecastRateIndexIDRef_ = forecastRateIndexNode.Attributes["href"].Value;
                    }
                    else
                    {
                        forecastRateIndex_ = new ForecastRateIndex(forecastRateIndexNode);
                    }
                }
                else
                {
                    forecastRateIndex_ = new ForecastRateIndex(forecastRateIndexNode);
                }
            }
            
        
        }
        
    
        #region Algorithm_
        private XsdTypeString algorithm_;
        public XsdTypeString Algorithm_
        {
            get
            {
                if (this.algorithm_ != null)
                {
                    return this.algorithm_; 
                }
                else if (this.algorithmIDRef_ != null)
                {
                    algorithm_ = IDManager.getID(algorithmIDRef_) as XsdTypeString;
                    return this.algorithm_; 
                }
                else
                {
                      return this.algorithm_; 
                }
            }
            set
            {
                if (this.algorithm_ != value)
                {
                    this.algorithm_ = value;
                }
            }
        }
        #endregion
        
        public string algorithmIDRef_ { get; set; }
        #region ForecastRateIndex_
        private ForecastRateIndex forecastRateIndex_;
        public ForecastRateIndex ForecastRateIndex_
        {
            get
            {
                if (this.forecastRateIndex_ != null)
                {
                    return this.forecastRateIndex_; 
                }
                else if (this.forecastRateIndexIDRef_ != null)
                {
                    forecastRateIndex_ = IDManager.getID(forecastRateIndexIDRef_) as ForecastRateIndex;
                    return this.forecastRateIndex_; 
                }
                else
                {
                      return this.forecastRateIndex_; 
                }
            }
            set
            {
                if (this.forecastRateIndex_ != value)
                {
                    this.forecastRateIndex_ = value;
                }
            }
        }
        #endregion
        
        public string forecastRateIndexIDRef_ { get; set; }
        
    
        
    
    }
    
}

